### ISSUE TITLE
Best practices for libvirt and zfs images

### ISSUE THREAD

**Issue by candlerb** [jimsalterjrs/sanoid#191]
First, an apology: this is really a request for information, rather than an issue report.  If there were a discussion group for sanoid I would have posted there instead.

I have watched the sanoid/syncoid videos and read the articles, and have been trying to piece together what you are running on top of them, to save reinventing the wheel for a robust setup.

1. Looking at the videos, I believe you are using the virt-manager GUI on top of libvirt, as your way of managing kvm instances. Correct?

2. I wondered how you were creating zfs datasets per VM, and in particular if you were using [libvirt's zfs support](https://libvirt.org/storage.html#StorageBackendZFS) to create the VM disk images.  So I tested under Ubuntu 16.04 by following [this blog post](http://empt1e.blogspot.co.uk/2014/08/zfs-support-in-libvirt.html) and as far as I can tell, what you end up with are zvols rather than image files:

~~~
# virsh vol-create-as --pool zfsimages --name vol1 --capacity 1G

# zfs list
NAME                    USED  AVAIL  REFER  MOUNTPOINT
brian-kit              1.03G   210G    96K  /brian-kit
brian-kit/images       1.03G   210G    96K  /brian-kit/images
brian-kit/images/vol1  1.03G   211G    64K  -

# virsh vol-list zfsimages
 Name                 Path
------------------------------------------------------------------------------
 vol1                 /dev/zvol/brian-kit/images/vol1

# ls -l /dev/zvol/brian-kit/images/
total 0
lrwxrwxrwx 1 root root 12 Jan  6 13:47 vol1 -> ../../../zd0
~~~

Since you wrote an article saying [don't use zvols](http://www.openoid.net/psa-snapshots-are-better-than-zvols/), then I'm guessing you *don't* use this.  Furthermore, virt-manager doesn't seem to allow creation of images in zfs pools; the plus button is dimmed out.

So, do you just manually create a zfs dataset at the command line, for each VM you create?  Or do you have another way to hook into virt-manager, to create a zfs dataset plus image file?

There's another issue in virt-manager.  Making a libvirt "dir" pool for `/brian-kit/images` doesn't help in the GUI, because at the point where you create a VM it doesn't let you put the file in a subdirectory, e.g. `test1/vda.img`.

So it looks like I have to create both the dataset and the image file by hand:

~~~
zfs create brian-kit/images/test1
truncate -s 10G /brian-kit/images/test1/vda.img
~~~

...and then use "browse local" to attach the file in virt-manager.  It's doable, but I wonder if I'm missing a cleaner way to do this.

3. Assuming you are using image files on a zfs dataset, are you using raw images or qcow2?

    It bothers me what might happen to a qcow2 image if you take a zfs snapshot of it mid-update, i.e. whether it could end up in an inconsistent state - so perhaps raw images are safer.  But qcow2 images do let you do things like snapshot the running VM state, including RAM and registers.

    If they are raw images, I presume they are sparse?  ISTM that since zfs never overwrites blocks in-place, there would be no benefit to pre-allocated zeroed storage.

4. Are you using zfs inside your guest VMs as well, or something traditional like ext4?  Are there any performance (or other) issues to beware of using zfs inside an image file which itself is stored on zfs?

5. When sanoid creates a snapshot of a dataset containing a VM image, does it quiesce the VM first?

    This would be to avoid the guest's filesystem being in an inconsistent state at the time the snapshot is taken.  Maybe this doesn't matter if the guest is using ZFS internally - it should always be consistent, and [ZoL doesn't support it yet](https://github.com/zfsonlinux/zfs/issues/260) anyway.

    References:

    * https://linux.die.net/man/8/fsfreeze
    * https://wiki.libvirt.org/page/Qemu_guest_agent
    * https://specs.openstack.org/openstack/nova-specs/specs/kilo/implemented/quiesced-image-snapshots-with-qemu-guest-agent.html
    * https://github.com/sheepdog/sheepdog/wiki/How-to-make-a-live-snapshot-properly

6. Any suggestions for managing the libvirt XML domain files? In one of the videos you show starting a remote clone - but presumably it was necessary to dump the XML from the local host and import it into the remote host first.  Or do you have a smarter way to manage this?

7. Any other hints for a robust libvirt + zfs setup? Is having sanlock or virtlockd useful? What about the different [disk cache modes](https://libvirt.org/formatdomain.html#elementsDisks) that qemu offers?

Maybe there's some source material for more articles here :-)

Thanks... Brian.

**Comment by samarium** [comment]
re 6: It is worth reading Jim's articles on openoid.com, especially this one http://www.openoid.net/reshuffling-pool-storage-on-the-fly/ as it shows how he syncs the vm confirguration etc by mounting zfs datasets at appropriate parts of the normal file system, so then the sync is just part another syncoid invocation.

I too would like to find someway to get virt-manager to handle sub-dirs better, and to alloc zfs dataset allocation for the vm, and file allocation inside of it.

I don't care about swap files and allocate in a separate dataset which I don't bother to snapshot or replicate.

**Comment by candlerb** [comment]
> re 6: It is worth reading Jim's articles on openoid.com, especially this one http://www.openoid.net/reshuffling-pool-storage-on-the-fly/ as it shows how he syncs the vm confirguration etc by mounting zfs datasets at appropriate parts of the normal file system

Ah thanks, I missed that bit: `/etc/libvirt/qemu` is itself mounted as a zfs filesystem.

**Comment by dabreese00** [comment]
I've been inspired and playing around with this recently too, and I've been wondering about many of the same questions, still trying to sort it out.  (I came across this while researching for ways to deal with the virt-manager/subdirectories issue.)

For now, here are a few bits I can try to contribute:

For one part of 3 and of 7, if I understand the question, I think Jim is using `.qcow2` with writeback caching.  If you haven't seen it already, he has an old-ish article with some relevant comments and benchmarks: http://jrs-s.net/2013/05/17/kvm-io-benchmarking/

Also towards 3: I think Jim (and others) advocate using compression on your ZFS datasets, which might make the whole question a wash either way?

In case it's helpful, as a ZFS newbie I've found the [ZoL Debian Stretch root how-to](https://github.com/zfsonlinux/zfs/wiki/Debian-Stretch-Root-on-ZFS) a great resource as well (I think they have these for a couple other major distros too).  Pinpointing system directories like `/etc/libvirt/qemu` to mount as individual ZFS datasets felt a bit hackish to me at first.  But not anymore, now that my whole box is ZFS; then it becomes very natural.

Hope this helps.  I, too, would be grateful to hear any attempt at elucidation on the rest of your questions.

**Comment by candlerb** [comment]
> I think Jim is using .qcow2 with writeback caching. If you haven't seen it already, he has an old-ish article with some relevant comments and benchmarks: http://jrs-s.net/2013/05/17/kvm-io-benchmarking/

Thanks for that pointer, which includes this sentence near the end: "In the real world, I use simple .qcow2 on ZFS, no zvols required."

> Also towards 3: I think Jim (and others) advocate using compression on your ZFS datasets, which might make the whole question of pre-allocation a wash either way?

Pre-allocation is mainly to avoid fragmentation of the underlying filesystem.  Jim does mention it when he says:

> I tested using fully-allocated qcow2 files.  I did a little bit of casual testing with sparsely allocated (aka “thin provisioned”) qcow2 files, and basically, they follow the same performance curves, but with roughly half the performance on some writes.  This isn’t  that big a deal, in my opinion – you only have to do a “slow” write to any given block once.  After that, it’s a rewrite, not a new write, and you’re back to the same performance level you’d have had if you’d fully allocated your qcow2 file to start with.

But that last sentence doesn't really apply to ZFS, since ZFS never does overwrites-in-place: new data is always written to empty blocks and the pointers updated.  Growing a qcow2 file or allocating blocks in a sparse file would do the same thing.

This hints that it's particularly beneficial to use SSD for a ZFS filesystem which holds VM images, to minimise the seek overheads.  As for compression: I'd say yes, just turn it on.  CPU overhead is low, and it reduces the size of I/O transfers.

**Comment by kviolka** [comment]
http://www.linuxsystems.it/2018/10/consistently-backup-your-virtual-machines-using-libvirt-and-zfs-part-1/

This sounds like the way to go!

**Comment by phreaker0** [comment]
Closing as the the required stuff (pre/post scripts #280, ...) are all in master

**Comment by mlsorensen** [comment]
So, I came across this in researching libvirt's zfs support, and I'd like to point something out about the "don't use zvols" mentioned. 

There's a misunderstanding in the linked article about zvols and snapshots, which seems to be the driver for recommending against zvols.

1) Not being able to snapshot zvols if you don't have enough space to rewrite the whole zvol is a safety measure, to ensure you don't get errors as you rewrite arbitrary blocks on the volume. It's a guarantee that the zvol can be totally rewritten. This is fundamentally different from a filesystem, where the usage is file by file within the particular filesystem. With sparse qcow2, one is free to shoot themselves in the foot with an ENOSPC when the zfs it is on runs out of storage.

2) This feature can be disabled. It happens because the zvol's default 'refreservation' is equal to the size of the zvol, which blocks any operations that would eat into that reserved space on the pool. Simply set the zvol's refreservation down, if you prefer to be able run your storage overallocated.

3) This behavior only occurs when space is actually being used. One is free to snapshot a zvol over and over and over if there's no actual data in the zvol, so even with the default behavior you do get some ability to store more than you'd expect. You only need to be able to store snapshotsize for each snapshot + zvol size.

In summary I don't see any deal breakers here as far as using zvols, if libvirt can natively manage them.

**Comment by tgunr** [comment]
One of the point about refreservation was it is something that has to be explicitly set, e.g. not a default. But I suppose that "could" be done via scripting.
